### PR TITLE
Farm and cage start dates functionality

### LINKED ISSUE #72: Farm and cage start days
At the moment the code does not take into account the start dates of farms and cages. 

Cages that have not started yet should not be taken into consideration for cage updates - since we assume it has not been populated with fish yet (so no treatment, fish mortality, lice progression dependent on fish, etc).

However, farms that have not started should still be available for interfarm travel as the farm might start operating before the eggs from different, already operating farm arrive at it.

### PR DESCRIPTION
**Pull request by SaraJakubiak**
Closes #72 

- Under the new changes non-operating farms and cages that has not yet reached starting date are updated but only with updates that do not require the fish
- Moved `LicePopulation` and queue batches from `Cage.py` to separate files
- Added wee fix to stop pytest numpy warnings

### CODE REVIEW

File: refactor/src/Cage.py [hunk 1]
```diff
@@ -242,17 +178,20 @@ def update(self, cur_date: dt.datetime, step_size: int, pressure: int) -> tuple:
...
             hatched_arrivals_dist
         )
 
-        self.logger.debug("\t\tfinal fish population = {}".format(self.num_fish))
         self.logger.debug("\t\tfinal lice population= {}".format(self.lice_population))
-        egg_distrib = new_egg_batch.geno_distrib
-        hatch_date = new_egg_batch.hatch_date
-        self.logger.debug("\t\tlice offspring = {}".format(sum(egg_distrib.values())))
+
+        egg_distrib = {}    # type: GenoDistrib
```
Comment (erolm-a) [inline refactor/src/Cage.py:188]: Note: this is actually a review for line 100, but there is something to say about this line as well.

Firstly: `tuple` is a type constructor, not a type annotation. I think the wished return type for `Cage.update()` is `Tuple[GenoDistrib, Optional[dt.datetime]]`. In lack of a proper type annotation pytype may give warnings or even errors. I believe if the proper return type is given there is no need to put a `# type: GenoDistrib` here. Similarly for the line below. I am not sure how this slipped from my last PR. My bad.

In general, I recommend running stuff through mypy as well. It is way noisier (hence why I did not include it in the CI) but it helps enforcing more compliance.

File: refactor/src/Cage.py [hunk 2]
```diff
@@ -679,26 +618,12 @@ def update_distrib_discrete_subtract(distrib_delta, distrib):
...
-            return dict(zip(keys, map(int, np.zeros_like(np_values))))
-        np_values = np_values * population / np.sum(np_values)
-        np_values = np_values.round().astype(np.int32)
-        # correct casting errors
-        np_values[-1] += population - np.sum(np_values)
-
-        return dict(zip(keys, map(int, np_values)))
-
-    def select_dams(self, distrib_dams_available: GenoDistrib, num_dams: int):
+    def select_dams(self, distrib_dams_available, num_dams):
```
Comment (erolm-a) [inline refactor/src/Cage.py:621]: Is there a reason why this type annotation got removed?

Comment (SaraJakubiak) [reply refactor/src/Cage.py:621]: Most likely a merge omission - will add it back

File: refactor/src/Cage.py [hunk 3]
```diff
@@ -474,7 +477,7 @@ def get_variance_infected_fish(self, n: int, k: int) -> float:
         probs = np.full(n, 1/n)
         bins = self.cfg.rng.multinomial(k, probs)
 
-        return np.var(bins) * k
+        return float(np.var(bins) * k)
```
Comment (erolm-a) [inline refactor/src/Cage.py:480]: We need to choose to which datatype to stick for scalar values. This applies both here and in a number of places (eg. the parametrers of update_deltas, why some are int64 and other are int?) and mypy is extremely unhappy about this.

File: refactor/src/Cage.py [hunk 4]
```diff
@@ -815,7 +818,7 @@ def update_deltas(
             new_males: np.int64,
```
Comment (erolm-a) [inline refactor/src/Cage.py:818]: why this is np.int64 and new_infections is int?

File: refactor/src/Cage.py [hunk 5]
```diff
@@ -815,7 +818,7 @@ def update_deltas(
...
             new_infections: int,
             lice_from_reservoir: dict,
```
Comment (erolm-a) [inline refactor/src/Cage.py:820]: Reminder to myself these dicts should be better typed. Also, `dict` is not a proper way to annotate!

File: refactor/src/Cage.py [hunk 6]
```diff
@@ -132,10 +135,19 @@ def update(self, cur_date: dt.datetime, step_size: int, pressure: int) -> tuple:
         # Lice coming from reservoir
         lice_from_reservoir = self.get_reservoir_lice(pressure)
 
-        if cur_date >= self.start_date:
-            # Events that only happen when cage is populated with fish
-            # (after start date)
+        if cur_date < self.start_date:
+            # Values that are not used before the start date
+            treatment_mortality = {}
+            fish_deaths_natural, fish_deaths_from_lice = 0, 0  # type: Tuple[int, int]
```
Comment (erolm-a) [inline refactor/src/Cage.py:141]: Why are we using these type annotations? I believe both mypy and pytype can easily tell the type of fish_deaths_natural and fish_deaths_from_lice. Unless it's related to int vs np.int64? (See below)

File: refactor/src/Cage.py [hunk 7]
```diff
@@ -132,10 +135,19 @@ def update(self, cur_date: dt.datetime, step_size: int, pressure: int) -> tuple:
...
+            num_infection_events = 0  # type: int
+            delta_avail_dams, delta_eggs = {}, {}  # type: Tuple[GenoDistrib, GenoDistrib]
+            avail_dams_batch = None  # type: OptionalDamBatch
```
Comment (erolm-a) [inline refactor/src/Cage.py:144]: Not needed: the return type allows for both pytype (and possibly mypy) to infer the type of avail_dams_batch.

Types are meant to be friendly and supportive, not a nuisance to carry around!

File: refactor/src/LicePopulation.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    @available_dams.setter
+    def available_dams(self, new_value: GenoDistrib):
+        for geno in new_value:
+            assert self.geno_by_lifestage["L5f"][geno] >= new_value[geno], \
+                f"current population geno {geno}:{self.geno_by_lifestage['L5f'][geno]} is smaller than new value geno {new_value[geno]}"
+
+        self._available_dams = new_value
+
+    @staticmethod
+    def multiply_distrib(distrib: dict, population: int):
```
Comment (erolm-a) [inline refactor/src/LicePopulation.py:60]: I was thinking, would it be a good idea to move the update_discrete_add/subtract functions here? They'd have to be replaced with overloaded definitions anyway?

File: refactor/tests/test_Cage.py [hunk 9]
```diff
@@ -619,3 +627,93 @@ def test_get_reservoir_lice(self, first_cage):
...
+        inflow = pressure + sum(geno_hatch.values())
+
+        # set mortality to 0
+        first_cage.cfg.background_lice_mortality_rates = {key: 0 for key in first_cage.lice_population}
+
+        offspring, hatch_date = first_cage.update(cur_date, 1, pressure)
+
+        assert offspring == {}
+        assert hatch_date is None
+        assert first_cage.lice_population["L3"] == 0
```
Comment (erolm-a) [inline refactor/tests/test_Cage.py:687]: What about this: assert `all(first_cage.lice_population[susceptible_stage] == 0 for susceptible_stage in Cage.susceptible_stages)`

File: refactor/src/Cage.py (hunk 3 above)
Comment (SaraJakubiak) [reply refactor/src/Cage.py:480]: I think this deserves separate issue - added #84

File: refactor/src/Cage.py (hunk 4 above)
Comment (SaraJakubiak) [reply refactor/src/Cage.py:818]: #84

File: refactor/src/Cage.py (hunk 7 above)
Comment (SaraJakubiak) [reply refactor/src/Cage.py:144]: For both this and previous comment - I went nuclear with annotation after mypy complained. Will revise and remove unneeded annotations.

**Review by erolm-a — APPROVED** [review]

**Pull request merged**